modules/wallet/scan.go

```diff
@@ -0,0 +1,198 @@
+package wallet
+
+import (
+	"fmt"
+
+	"github.com/NebulousLabs/Sia/build"
+	"github.com/NebulousLabs/Sia/modules"
+	"github.com/NebulousLabs/Sia/types"
+)
+
+// numInitialKeys is the number of keys generated by the seedScanner before
+// scanning the blockchain for the first time.
+var numInitialKeys = func() uint64 {
+	switch build.Release {
+	case "dev":
+		return 1e4
+	case "standard":
+		return 1e6
+	case "testing":
+		return 1e3
+	default:
+		panic("unrecognized build.Release")
+	}
+}()
+
+// maxScanKeys is the number of maximum number of keys the seedScanner will
+// generate before giving up.
+var maxScanKeys = func() uint64 {
+	switch build.Release {
+	case "dev":
+		return 1e6
+	case "standard":
+		return 100e6
+	case "testing":
+		return 1e5
+	default:
+		panic("unrecognized build.Release")
+	}
+}()
+
+var errMaxKeys = fmt.Errorf("refused to generate more than %v keys from seed", maxScanKeys)
+
+type scannedOutput struct {
+	id        types.OutputID
+	value     types.Currency
+	seedIndex uint64
+}
+
+// A seedScanner scans the blockchain for addresses that belong to a given
+// seed.
+type seedScanner struct {
+	blockheight      types.BlockHeight
+	dustThreshold    types.Currency              // minimum value of outputs to be included
+	keys             map[types.UnlockHash]uint64 // map address to seed index
```

Comment:
haven't looked at much of the rest of the code, but is it necessary to store the whole unlock hash when doing the scan? That's going to add something like 60MB to the memory-cost of this operation for 1e6 seeds, and 6GB of memory cost for 100e6 seeds.

Because the settings go all the way up to 100e6 seeds, the memory usage needs to be thought through a bit more. Even at 1e6 seeds it's going to be an expensive operation.